### ISSUE TITLE
ngOnInit not always called if OnPush strategy is used

### ISSUE THREAD

**Issue by cartant** [angular/angular#9976]
**I'm submitting a ...**
```
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

If `ChangeDetectionStrategy.OnPush` is used for a component that has a `router-outlet`, a component placed into the outlet does not always have its `ngOnInit` function called.

If `ChangeDetectionStrategy.Default` is used, the `ngOnInit` function is always called.

**Expected/desired behavior**

The `ngOnInit` function should always be called be called, regardless of the change detection strategy.

**Reproduction of the problem**

There is a plunker [here](http://plnkr.co/edit/KuJsrlUlD61s9JDEyee4) to demonstrate the problem.

There seem to be two instances in which `ngOnInit` is not called:

* for the initial component placed into the outlet; and
* when using the browser's navigation buttons.

If you open the console and run the plunker, you should see that `ngOnInit` is not called for the first component. If you click the "second" button to navigate to the second component, you should see that `ngOnDestroy` is called on the first component and `ngOnInit` is called on the second component. Thereafter, navigating between the two results in the expected sequence of `ngOnDestroy/Init` calls.

However, if you navigate using the browser's back and next buttons, you should see that `ngOnInit` is not called.

If you stop and re-run the plunker and then click the "greet" button, you should see that `ngOnInit` is called on the first component when an `ngIf` within it is evaluated.

If you stop the plunker and set the app component's change detection strategy to `Default`, `ngOnInit` should be always be called - even when navigating using the browser's back and next buttons.

**What is the expected behavior?**

As above: the `ngOnInit` function should always be called be called.

**Related/possibly-related issues**

There are a number of issues that seem to relate to `ngOnInit` not being called. The first of these might [reference a fix for this issue](https://github.com/angular/angular/issues/9818#issuecomment-230808580), but I'm unsure of the relationship between change detection and life-cycle management:

* [router-outlet not triggering change-detection after component-insertion](https://github.com/angular/angular/issues/9818)

And there are others that seem to relate to `ngOnInit` and browser-based navigation:

* [Bug: OnInit not fired when using Safari's back button](https://github.com/angular/angular/issues/7722)
* [OnInit is not called when I type URL in IE ](https://github.com/angular/angular/issues/9860)

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.4

* **Angular router:** 3.0.0-beta.2
 
* **Browser:** Chrome 51 
 
* **Language:** TypeScript 1.8

**Comment by wardbell** [comment]
Misbehavior confirmed for app using the v.3 router in IE 11. Try running the Documentation router demo. Best to run `npm start` after set-up. 

You can also see it [in plunker](https://angular.io/resources/live-examples/router/ts/plnkr.html)
1) Launch IE 11 (supposedly also fails in Safari)
2) Set address bar to: https://angular.io/resources/live-examples/router/ts/plnkr.html
3) It displays Crisis Center
4) Click Heroes link
5) Press [backspace] key (the browser [back] equivalent).
6) Observe that the Crisis Center component is displayed but none of its contents ... because `ngOnInit` not called.
7) Click the Crisis Center link; now it works.

Rumor has it that the problem has been traced to zones and that a zones hack can "fix" it.

**Comment by arunsn43** [comment]
I am still getting this behavior from RC 7.  ChangeDetectionStrategy is set to OnPush. 
ngOnInit is not called when the module is lazily loaded or through dynamic resolution.  
How to fix this?

**Comment by cartant** [comment]
@arunsn43 You might want to open another issue, as I've just noticed that this issue appears to have been fixed.

The Plunker I created for this issue was a bit of a mess, as I didn't pin the versions in the SystemJS configuration. I've pinned it and I've created another for 2.0.0-rc.7 and another for 2.0.1:

* [Plunker using 2.0.0-rc.4](http://plnkr.co/edit/KuJsrlUlD61s9JDEyee4)
* [Plunker using 2.0.0-rc.7](https://plnkr.co/edit/hIYnkPrKf70SpjHwZUIW)
* [Plunker using 2.0.1](https://plnkr.co/edit/Ewb9qHGPt6B3CK7eUaAC)

The issue isn't reproducible in 2.0.0-rc.7 nor in 2.0.1, so I'm closing this. If you still have a problem, it likely  differs from the one that prompted this issue.
